### PR TITLE
Add new QR reader widget support

### PR DESCRIPTION
**Pull request by LZRS**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes https://github.com/opensrp/fhircore/issues/3350

**Description**

  - Overrides barcode layout to contain required missing views
  to avoid crash on render with error
  
      ```
      FATAL EXCEPTION: main
                                                                                                          Process: org.smartregister.opensrp.diabetesCompassClinic, PID: 8413
                                                                                                          java.lang.NullPointerException: findViewById(...) must not be null
                                                                                                    	      at com.google.android.fhir.datacapture.views.factories.QuestionnaireItemViewHolder.<init>(QuestionnaireItemViewHolderFactory.kt:64)
                                                                                                    	      at com.google.android.fhir.datacapture.views.factories.QuestionnaireItemViewHolderFactory.create(QuestionnaireItemViewHolderFactory.kt:37)
                                                                                                    	      at com.google.android.fhir.datacapture.QuestionnaireEditAdapter.onCreateViewHolderQuestion(QuestionnaireEditAdapter.kt:88)
                                                                                                    	      at com.google.android.fhir.datacapture.QuestionnaireEditAdapter.onCreateViewHolder(QuestionnaireEditAdapter.kt:64)
                                                                                                    	      at com.google.android.fhir.datacapture.QuestionnaireEditAdapter.onCreateViewHolder(QuestionnaireEditAdapter.kt:49)
                                                                                                    	      at androidx.recyclerview.widget.RecyclerView$Adapter.createViewHolder(RecyclerView.java:7295)
      ```

**Engineer Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [ ] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

**Review by ellykits — DISMISSED** [review]

**Comment by pld** [comment]
@LZRS looks good! seems like the only thing missing is test coverage

**Comment by pld** [comment]
>  

Good catch! That string does not look correct, we also want to set the string from fhircore because we need to be able to translate it

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterViewModel.kt [hunk 1]
```diff
@@ -166,10 +166,10 @@ constructor(
     when (event) {
       // Search using name or patient logicalId or identifier. Modify to add more search params
       is RegisterEvent.SearchRegister -> {
-        if (event.searchText.isEmpty()) {
+        if (event.searchQuery.isBlank()) {
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterViewModel.kt:169]: We need both checks here. You can create an extension in the `StringExtension.kt` file `isEmptyOrBlank()`

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.quest.ui.shared.models
+
+import org.smartregister.fhircore.engine.configuration.workflow.ActionTrigger
+
+sealed class UiSearchMode {
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt:21]:
```suggestion
sealed class SearchMode {
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * Wrapper class to hold search input [String] and the [UiSearchMode] mode used to initiate the UI
+ * search
+ *
+ * Depending on the [UiSearchMode], additional [ActionTrigger.ON_SEARCH_SINGLE_RESULT] actions can
+ * be triggered a query returns a single result
+ *
+ * @param query Actual search input string
+ * @param mode [UiSearchMode] that initiated the search
+ */
+data class UiSearchQuery(val query: String, val mode: UiSearchMode = UiSearchMode.KeyboardInput) {
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt:37]:
```suggestion
data class SearchQuery(val query: String, val mode: SearchMode = UiSearchMode.KeyboardInput) {
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  fun isEmpty() = query.isEmpty()
+
+  fun isBlank() = query.isBlank()
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt:40]: StringExtensions#isEmptyOrBlank

Then invoke the function here.

```kotlin 
fun isEmptyOrBlank() =  query.isEmpty() || query.isBlank()
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensions.kt [hunk 5]
```diff
@@ -74,151 +74,176 @@ fun List<ActionConfig>.handleClickEvent(
...
+      )
+    }
+    ?.let { theConfig ->
+      val computedValuesMap = resourceData?.computedValuesMap ?: emptyMap()
+      val actionConfig = theConfig.interpolate(computedValuesMap)
+      val interpolatedParams = interpolateActionParamsValue(actionConfig, resourceData)
+      val practitionerId =
+        interpolatedParams
+          .find { it.paramType == ActionParameterType.RESOURCE_ID && it.key == PRACTITIONER_ID }
+          ?.value
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensions.kt:108]: Can you create a different action parameter type for PRACTITIONER_ID instead of reusing the RESOURCE_ID? the resource ID was meant to be used differently. Only the parameter type should be required for an action parameter. The keys are dynamic.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensions.kt [hunk 6]
```diff
@@ -74,151 +74,176 @@ fun List<ActionConfig>.handleClickEvent(
...
+  navController: NavController,
+  resourceData: ResourceData? = null,
+  navMenu: NavigationMenuConfig? = null,
+  context: Context? = null,
+) {
+  this.takeIf {
+      it.trigger.isIn(
+        ActionTrigger.ON_SEARCH_SINGLE_RESULT,
+        ActionTrigger.ON_CLICK,
+        ActionTrigger.ON_QUESTIONNAIRE_SUBMISSION,
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensions.kt:98]: This will not be necessary. The function `firstOrNull`/`find` will apply the filter based on the predicate provided.

File: android/quest/src/test/java/org/smartregister/fhircore/quest/util/QrCodeScanUtilsTest.kt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+class QrCodeScanUtilsTest : RobolectricTest() {
+  @get:Rule(order = 0) var hiltAndroidRule = HiltAndroidRule(this)
+
+  @Before
+  fun setUp() {
+    hiltAndroidRule.inject()
+  }
+
+  @Test
+  fun scanQrCodeShouldInvokeOnQrCodeResultWithTheQrCode() {
```
Comment (ellykits) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/util/QrCodeScanUtilsTest.kt:47]: Would it be possible to refactor this test by reducing the number of mocks or avoiding mocks in totality?

File: android/quest/src/test/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensionsKtTest.kt [hunk 8]
```diff
@@ -464,8 +462,8 @@ class ConfigExtensionsTest : RobolectricTest() {
     verify {
       context.startActivity(
         withArg {
-          assertEquals(it.action, Intent.ACTION_DIAL)
-          assertEquals(it.data, Uri.parse("tel:0700000000"))
+          kotlin.test.assertEquals(it.action, Intent.ACTION_DIAL)
+          kotlin.test.assertEquals(it.data, Uri.parse("tel:0700000000"))
```
Comment (ellykits) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensionsKtTest.kt:466]: Kindly refactor to use JUnit

File: android/quest/src/test/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensionsKtTest.kt [hunk 9]
```diff
@@ -578,9 +576,9 @@ class ConfigExtensionsTest : RobolectricTest() {
           ),
       )
     val resultOfInterpolatedValues = interpolateActionParamsValue(actionConfig, resourceData)
-    assertEquals(4, resultOfInterpolatedValues.size)
-    assertEquals("param2", resultOfInterpolatedValues[1].key)
-    assertEquals("1235", resultOfInterpolatedValues[1].value)
+    kotlin.test.assertEquals(4, resultOfInterpolatedValues.size)
+    kotlin.test.assertEquals("param2", resultOfInterpolatedValues[1].key)
+    kotlin.test.assertEquals("1235", resultOfInterpolatedValues[1].value)
```
Comment (ellykits) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensionsKtTest.kt:581]: Use JUnit. That's the standard for FHIRCore at the moment.

File: android/quest/src/test/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensionsKtTest.kt [hunk 10]
```diff
@@ -685,8 +683,8 @@ class ConfigExtensionsTest : RobolectricTest() {
           decodedImageMap,
         )
       }
-      assertTrue(decodedImageMap.isNotEmpty())
-      assertTrue(decodedImageMap.containsKey("d60ff460-7671-466a-93f4-c93a2ebf2077"))
+      TestCase.assertTrue(decodedImageMap.isNotEmpty())
+      TestCase.assertTrue(decodedImageMap.containsKey("d60ff460-7671-466a-93f4-c93a2ebf2077"))
```
Comment (ellykits) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensionsKtTest.kt:687]: Use JUnit. That's the standard for FHIRCore at the moment.

Comment (ellykits) [reply android/quest/src/test/java/org/smartregister/fhircore/quest/util/extensions/ConfigExtensionsKtTest.kt:687]: Should be applicable to the rest of the tests.

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt (hunk 4 above)
Comment (ellykits) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/shared/models/Search.kt]: It does. However, it throws a fatal `IllegalArgumentException` if the string is either blank or empty, that should be handled.

**Review by ellykits — APPROVED** [review]
